### ISSUE TITLE
[bug] conan.tools.scm.Git.commit_in_remote returns fails for tags in a multibranch Jenkins pipeline

### ISSUE THREAD

**Issue by puetzk** [conan-io/conan#14159]
### Environment details

* Operating System+version: ubuntu 22.04
* Conan version: 1.60.1
* Python version: 3.10.6
* Jenkins version: Jenkins 2.387.2

### Steps to reproduce

1. Use the normal [github-branch-source](https://plugins.jenkins.io/github-branch-source/) plugin, enabling `Discover tags` so it indexes tags as well as branches/PRs.
2. Have a tag containing a default `checkout scm` step
3. Use `conan export` on a recipe that follows the guidelines of https://docs.conan.io/2/examples/tools/scm/git/capture_scm/git_capture_scm.html?highlight=export%20scm

Result: 
Branches and PRs will use the public URL in their conandata.yml, but tags will instead use the local filesystem path (to a jenkins workspace on the agent). 

The problem seems to be the refspecs used by Jenkins vs the logic of commit_in_remote:
https://github.com/conan-io/conan/blob/81b195262350a13525a44a0b079c9a4eab9fb1cd/conan/tools/scm/git.py#L81

`git branch -r --contains`, is basically searching the ancestry of local refs named `refs/remotes/*`. I would propose that this should also consider `git tag --contains <hash>` an acceptable match, on the basis that tags are a pretty clear indication of "intent to publish this commit", even though they are a global namespae rather than 1:1 associated with a particular URL like a remote-tracking branch is.

If really necessary, one could use `git ls-remote {remote} {tag}` to confirm that a particular tag was present on that remote, but that's a network roundtrip. I think it's sane enough to just assume that tags will eventually end up being pushed; if this exports a conandata.yml referring to a tag that isn't pushed to a particular remote yet, that's easily fixed (by pushing it) and in any case less broken than an uploaded recipe containing a local filesystem link (which is a recipe revision that could never work anywhere else).

### Logs

the `.git/config` generated by jenkins uses refspecs that look like:

For a branch:
```
[remote "origin"]
        url = https://github.com/org/repo.git
        fetch = +refs/heads/master:refs/remotes/origin/master
```

For a PR:
```
[remote "origin"]
        url = https://github.com/org/repo.git
        fetch = +refs/pull/74/head:refs/remotes/origin/PR-74
        fetch = +refs/heads/master:refs/remotes/origin/master
```

For a tag
```
[remote "origin"]
        url = https://github.com/org/repo.git
        fetch = +refs/tags/v0.14.0-rc.2:refs/tags/v0.14.0-rc.2
```

So the issue is that Jenkins, like a default `git clone`, fetches tags into refs/tags/* (a shared global namespace) rather than displacing them per-remote into refs/remotes/*.

**Comment by memsharded** [comment]
Thanks @puetzk for the suggestion.

I understand that ``Git.commit_in_remote`` is not an issue per se, and we could add something like ``Git.commit_is_tag`` that implements your check. And the real issue is that the higher level ``get_url_and_commit`` is returning the local folder instead of the URL because it is not accepting the tag as valid. 

If that is the case, maybe implementing the low-level ``Git.commit_is_tag`` is the right approach, and then doing the extra check in ``get_url_and_commit`` would be doable, and if necessary to control the high level behavior, we could even condition it to a ``conf`` (not necessary at this moment, just for the future).

However there is still something I'd like to understand better, about how Jenkins work and the expectations here:

- The ``get_url_and_commit`` will store the ``commit`` in the conandata, so the idea is that later a ``git checkout <commit>`` will happen.
- Is that ``commit`` in the remote or is it an artificial commit introduced by Jenkins? If it is artificial, then wouldn't it fail too in the future when building from source, even if the url is correct now, the checkout will fail? Because the method is not capturing the tag, only the commit.
- If the ``commit`` is indeed existing in the remote exactly, how is that it is not listed in the git ``/refs/remotes``?
- How would the ``git ls-remote {remote} {tag}`` help if the ``commit`` we are trying to checkout is not in the remote?


One of my concerns here is that I have seen many times usage of tags that are later not pushed to the remote, so the assumption that a commit will be uploaded in the future might be a bit fragile. Is it basically the assumption that "if it is tag this commit that doesn't exist in the remote, will exist in the remote because I will push it"? If that is the case, maybe it should be an opt-in?

**Comment by puetzk** [comment]
> I understand that Git.commit_in_remote is not an issue per se, and we could add something like Git.commit_is_tag that implements your check. And the real issue is that the higher level get_url_and_commit is returning the local folder instead of the URL because it is not accepting the tag as valid.

Correct. I wouldn't really care about commit_in_remote except that get_url_and_commit uses it (for reasons that are sensible, but in this case problematic ).

> Is that commit in the remote

yes

> If the commit is indeed existing in the remote exactly, how is that it is not listed in the git /refs/remotes?

In this case, because Jenkins only clones refs relevant to what it was going to build. So branches fetch that one branch, PRs fetch their source and target branch, and tags fetch the tag but no branches at all. This is to avoid downloading the whole history of a large/old repo (or a big monorepo, or whatever).

But it would also be possible that the commit is available on the remote in a tag, without being in any branch. (e.g. if a bugfix was cherry-picked back from master into some 1.x-maintenance branch, tagged as x.y.z+1, and after the release the branch is closed or rebased (leaving only the tag as a permanent ref). Conan's own repo has a one such at 0.26.2 you look at `git log --tags --not --remotes`. It would have a lot more except you've kept open branches for most of the old releases. Some projects clean them up, though that isn't the root cause here. A tag ref is enough for the remote to have the hash available, and Jenkins would not have fetched branches for a tag job. So both relevant facts are the same (the remote has the hash, and the local clone has no tracking branches named refs/remotes/*) regardless of whether a project keeps the release branches or prunes them.

> How would the `git ls-remote {remote} {tag}`

It's the way to actually check what refs the remote has (if you're online, which CI obviously is but conan in general might not be).

> I have seen many times usage of tags that are later not pushed to the remote, so the assumption that a commit will be uploaded in the future might be a bit fragile

Yeah. It is a an assumption. I don't think its any all that much more fragile than trusting the local refs (which could be out of date), but either could fail to match the actual current state of the remote.

**Comment by puetzk** [comment]
In conan2, this is much improved by https://github.com/conan-io/conan/pull/15023, which uses `git {remote} --refetch --dry-run {commit}` to double-check that the commit actually isn't available, rather than assuming it isn't just because there's no remote-tracking branch info (in both this case and the cited bug #14928, the motivation is a shallow clone from a tag).

Duplicate of #14928
